### PR TITLE
Avoid AIOOB in StatsCPUCollector

### LINKED ISSUE #4602: ArrayIndexOutOfBoundException in StatsCPUCollector
Light on details here, just a stack trace. Appears to have happened post-startup, did not bring down the server or otherwise apparently affect usability.

```
i.d.s.StatsCPUCollector    | got an exception reading /proc/stat:
java.lang.ArrayIndexOutOfBoundException: Index 4096 out of bounds for length 4096
    at io.deephaven.stats.StatsCPUCollector.skipNextField(StatsCPUCollector.java:166)
    at io.deephaven.stats.StatsCPUCollector.updateSys(StatsCPUCollector.java:273)
    at io.deephaven.stats.StatsCPUCollector.update(StatsCPUCollector.java:458)
    at io.deephaven.stats.StatsDriver.timedOut(StatsDriver.java:176)
    at io.deephaven.io.sched.YASchedulerImpl.dispatch(YASchedulerImpl.java:692)
    at io.deephaven.io.sched.YASchedulerImpl.work(YASchedulerImpl.java:766)
    at io.deephaven.net.impl.nio.FastNIODriver.run(FastNIODriver.java:204)
    at java.lang.Thread.run(Thread.java:829)
```

I believe this is running on WSL2 on windows, started from the python embedded server.

### PR DESCRIPTION
**Pull request by niloc132**
Fixes #4602
Co-authored-by: Robb Camera <robbcamera@illumon.com>

### CODE REVIEW

**Comment by niloc132** [comment]
Intellij informed me there was some code duplication that i could address, but the initial commit here purely cherry-picks the fix that Robb initially wrote. The second commit just applies spotless formatting.

**Review by rcaudy — COMMENTED** [review]
.

**Review by devinrsmith — APPROVED** [review]
I'm happy to move forward with this as-is; if we think there is value spending more time in this code, we can prioritize as necessary.

File: Stats/src/main/java/io/deephaven/stats/StatsCPUCollector.java [hunk 1]
```diff
@@ -216,10 +247,26 @@ private void updateSys() {
...
+                while (true) {
+                    statFile.getChannel().position(0);
+                    nb = readToSizedBuffer(statFile, statBuffer);
+
+                    if (nb == statBuffer.length) {
+                        // allocate larger read-buffer, and try again
+                        statBuffer = new byte[statBuffer.length * 2];
+
+                        statFile.close();
+                        statFile = new FileInputStream(PROC_STAT_PSEUDOFILE);
```
[the remark this answers is not in the document]
Comment (niloc132) [reply Stats/src/main/java/io/deephaven/stats/StatsCPUCollector.java:263]: size() is always zero

File: Stats/src/main/java/io/deephaven/stats/StatsCPUCollector.java [hunk 2]
```diff
@@ -216,10 +247,26 @@ private void updateSys() {
...
-                    statFile = new FileInputStream("/proc/stat");
+                    statFile = new FileInputStream(PROC_STAT_PSEUDOFILE);
+                }
+
+                int nb;
+                while (true) {
+                    statFile.getChannel().position(0);
+                    nb = readToSizedBuffer(statFile, statBuffer);
+
+                    if (nb == statBuffer.length) {
```
[the remark this answers is not in the document]
Comment (niloc132) [reply Stats/src/main/java/io/deephaven/stats/StatsCPUCollector.java]: Rewritten now, no more re-reading from start when resizing.

File: Stats/src/main/java/io/deephaven/stats/StatsCPUCollector.java [hunk 3]
```diff
@@ -203,7 +208,33 @@ private long getNextFieldLong(int nb) {
...
+    }
+
+    /**
+     * Attempt to read the entire contents of an already-opened file to a pre-defined buffer. If the buffer is not large
+     * enough for the entire contents of the file, then the caller should re-allocate the buffer, and try again
+     *
+     * @param inFile a file whose bytes we intend to read
+     * @param buffer an array of bytes, which will be populates with the contents of the file
+     *
+     * @return the number of bytes successfully read into the buffer
```
[the remark this answers is not in the document]
Comment (niloc132) [reply Stats/src/main/java/io/deephaven/stats/StatsCPUCollector.java]: No longer necessary, readToBuffer always reads the full contents.

File: Stats/src/main/java/io/deephaven/stats/StatsCPUCollector.java [hunk 4]
```diff
@@ -536,7 +517,7 @@ private void updatePerThreadCPU() {
      */
     public void update() {
         if (statBuffer == null) {
-            statBuffer = new byte[4096];
+            statBuffer = ByteBuffer.allocate(1);
```
Comment (rcaudy) [inline Stats/src/main/java/io/deephaven/stats/StatsCPUCollector.java]: Revert testing change.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**